physical/raft/fsm.go

```diff
@@ -243,6 +252,105 @@ func writeSnapshotMetaToDB(metadata *raft.SnapshotMeta, db *bolt.DB) error {
 	return nil
 }
 
+func (f *FSM) localNodeConfig() (*LocalNodeConfigValue, error) {
+	var configBytes []byte
+	if err := f.db.View(func(tx *bolt.Tx) error {
+		value := tx.Bucket(configBucketName).Get(localNodeConfigKey)
+		if value != nil {
+			configBytes = make([]byte, len(value))
+			copy(configBytes, value)
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	if configBytes == nil {
+		return nil, nil
+	}
+
+	var lnConfig LocalNodeConfigValue
+	if configBytes != nil {
+		err := proto.Unmarshal(configBytes, &lnConfig)
+		if err != nil {
+			return nil, err
+		}
+		f.desiredSuffrage = lnConfig.DesiredSuffrage
+		return &lnConfig, nil
+	}
+
+	return nil, nil
+}
+
+func (f *FSM) upgradeLocalNodeConfig() error {
+	// Read the local node config
+	lnConfig, err := f.localNodeConfig()
+	if err != nil {
+		return err
+	}
+
+	// Entry is already present. Get the suffrage value.
+	if lnConfig != nil {
+		f.desiredSuffrage = lnConfig.DesiredSuffrage
+		return nil
+	}
+
+	//
+	// This is the upgrade case where there is no entry
+	//
+
+	lnConfig = &LocalNodeConfigValue{}
+
+	// Refer to the persisted latest raft config
+	config := f.latestConfig.Load().(*ConfigurationValue)
+
+	// If there is no config, then this is a fresh node coming up. This could end up
+	// being a voter or non-voter. But by default assume that this is a voter. It
+	// will be changed if this node joins the cluster as a non-voter.
+	if config == nil {
+		lnConfig.DesiredSuffrage = f.desiredSuffrage
+		return f.persistDesiredSuffrage(lnConfig)
+	}
+
+	// Get the last known suffrage of the node and assume that it is the desired
+	// suffrage. There is no better alternative here.
+	for _, srv := range config.Servers {
+		if srv.Id == f.localID {
+			switch srv.Suffrage {
+			case int32(raft.Nonvoter):
+				lnConfig.DesiredSuffrage = "non-voter"
+			default:
+				lnConfig.DesiredSuffrage = "voter"
+			}
+		}
+		// Bring the intent to the fsm instance.
+		f.desiredSuffrage = lnConfig.DesiredSuffrage
+	}
+
+	return f.persistDesiredSuffrage(lnConfig)
+}
+
+// recordSuffrage is called when a node successfully joins the cluster. This
+// intent should land in the stored configuration. If the config isn't available
+// yet, we still go ahead and store the intent in the fsm. During the next
+// update to the configuration, this intent will be persisted.
+func (f *FSM) recordSuffrage(desiredSuffrage string) error {
```

Comment:
Do we need to hold the fsm's Lock in this function?

Reply:
Its not particularly contentious, in the sense that we should only touch the `desiredSuffrage` field once during startup, once when a node joins the cluster, and potentially in future whenever the suffrage of the node change (if and when we support it). Having said that it shouldn't hurt if we grab the lock either. Fixed.